README.md
# dialog.js
移动端Dialog弹出对话框插件，同时支持jQuery和Zepto

# 特性：
+ 支持四种类型的弹窗展示：alert、confirm、info、tips；
+ 方便定制，可以自定义按钮的文字、样式、回调函数；
+ 支持 jQuery 和 Zepto JavaScript工具库；
+ 4个状态事件，用来监听状态改变；对话框默认样式主要针对移动端展示做优化。

# DEMO：
查看效果，请点击→：[demo示例](http://sufangyu.github.io/project/dialog/)

![项目地址二维码](http://7xp00j.com1.z0.glb.clouddn.com/erweima-dialog.js.png)


# 三种类型的对话框：
**1）alert 类型；**
![alert类型](http://7xp00j.com1.z0.glb.clouddn.com/static/images/dialog-plugin-01.jpg)

**2）confirm 类型**
![confirm类型](http://7xp00j.com1.z0.glb.clouddn.com/static/images/dialog-plugin-02.jpg)

**3）info 类型；**
![info类型](http://7xp00j.com1.z0.glb.clouddn.com/static/images/dialog-plugin-03.jpg)


# 引用文件：

**1、CSS**
``` css
<link rel="stylesheet" href="css/dialog.css">
```
**2、JS**
``` javascript
<script src="js/zepto.min.js"></script>
<script src="js/dialog.js"></script>
```
**3、JS 调用**
``` javascript
$.dialog({
    contentHtml : '弹出对话框的内容部分'
});
 ```

 <br />
 
# 参数
| 选项            | 类型    |  默认值  |  说明  |
| :--------       | :-----  | :----    | :----  |
| type            | string  | 'alert'  |  对话框的种类（alert-确定； confirm-确定/取消； info）|
|titleText 	      | string  | '信息提示' | 标题文字|
|showTitle 	      | bool    | true       | 是否显示标题。值为：true、false|
|contentHtml      | string  | ''         | 主体部分的HTML内容。默认值为空|
|dialogClass      | string  | ''         | 对话框自定义的 class。默认值为空|
|autoClose 	      | number  | 0          | 自动关闭的延时时间（单位：毫秒）。为0时表示不自动关闭；大于0时，表示自动关闭对话框|
|overlayClose     | bool    | fasle      | 是否可点击遮罩层关闭对话框。值为：true、false|
|buttonText       | string  | { ok : '确定', cancel : '取消' } | 按钮的文字|
|buttonClass      | string  | {ok : '',  cancel : ''}          | 按钮的自定义 class|
|infoText         | string  | ''            |info 类型对话框的提示文字|
|infoIcon         | string  | ''            |info 类型对话框的提示图标。值为图标的路径|
|onClickOk        | fn      | function(){} 	| 回调函数，点击“确定”按钮时执行|
|onClickCancel    | fn      | function(){}  | 回调函数，点击“取消”按钮时执行|
|onBeforeShow     | fn      | function(){}  | 回调函数，对话框显示前时执行|
|onShow 	      | fn      | function(){} 	| 回调函数，对话框显示后时执行|
|onBeforeClosed   | fn      | function(){}  | 回调函数，对话框关闭前时执行|
|onClosed 	      | fn      | function(){} 	| 回调函数，对话框关闭后时执行|


# 方法
| 方法            |   说明  |
| :--------       | :----   |
| $.dialog.close  | 关闭对话框。用法：dialogObj.dialog.close() |
| $.dialog.update | 更新info、tips类型对话框的状态（图标、提示文字），该方法可传入autoClose、infoText、infoIcon三个参数。<br />dialogObj.dialog.update({ autoClose : 1500, infoText : '操作成功', infoIcon : 'images/icon/success.png' })


#最新版本
### 1.0.3 (2016-07-15)
+ 修复弹窗在低版本安卓手机无法居中显示；
+ 修复连续点击遮罩层后，弹窗会自动关闭；
+ 延迟阻止Tap点透层的关闭时间；
+ 增加 info 和 tips 类型弹窗的关闭前与关闭后的回调函数；
+ 通过CSS3启用GPU硬件加速

[全部更新日志](https://github.com/sufangyu/dialog.js/blob/master/Changelog.md)


Changelog.md
###1.0.3 (2016-07-15)
+ 修复弹窗在低版本安卓手机无法居中显示；
+ 修复连续点击遮罩层后，弹窗会自动关闭；
+ 延迟阻止Tap点透层的关闭时间；
+ 增加 info 和 tips 类型弹窗的关闭前与关闭后的回调函数；
+ 通过CSS3启用GPU硬件加速

###1.0.2 (2016-06-22)
+ 修复弹出层按钮点击tap点透问题；
+ 修复使用jQuery时，tap无法正常使用；
+ 修复使用Zepto时，弹出层无法正常限制最大高度；
+ 优化弹出层的显示样式。

###1.0.1 (2016-06-19)
+ 修改弹出层按钮点击事件为自定义tap事件（PC使用click，Mibile使用tap）；
-增加弹出层主体内容不超出可视窗口高度限制；
+ 增加弹出层显示时，底层body内容不能触控滚动。

###1.0.0 (2016-06-04)
+ 支持四种类型的弹窗展示：alert、confirm、info、tips；
+ 方便定制，可以自定义按钮的文字、样式、回调函数；
+ 支持 jQuery 和 Zepto JavaScript工具库。


js/example.js
$(document).ready(function(){

	/*  
	 * alert type 
	 */
	$('#btn-01').click(function(){
		$.dialog({
			contentHtml : '<p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。end~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p><p>我是默认弹出对话框示例展示。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。end~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</p>'
		});
	});

	$('#btn-02').click(function(){
		$.dialog({
			autoClose : 2500,
			contentHtml : '<p>我是自动关闭的对话框示例展示。</p><p>我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p>'
		});
	});


	$('#btn-03').click(function(){
		$.dialog({
			showTitle : false,
			contentHtml : '<p>我是没有标题的对话框。</p><p>我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p>'
		});
	});


	$('#btn-04').click(function(){
		$.dialog({
			titleText : '自定义标题',
			contentHtml : '<p>我是自定义标题的对话框。</p><p>我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p>'
		});
	});


	/*  
	 * confirm type 
	 */
	$('#btn-05').click(function(){
		$.dialog({
			type : 'confirm',
			contentHtml : '<p>我是confirm类型的对话框。</p><p>我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p>'
		});
	});

	$('#btn-06').click(function(){
		$.dialog({
			type : 'confirm',
			buttonText : {
	            ok : '自定义-确定',
	            cancel : '自定义-取消'
	        },
			contentHtml : '<p>我是confirm类型的对话框。</p><p>我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p>'
		});
	});

	$('#btn-07').click(function(){
		$.dialog({
			type : 'confirm',
			onClickOk : function(){
				alert('你点了确定~~');
			},
        	onClickCancel : function(){        		
				alert('你点了取消~~');
        	},
			contentHtml : '<p>我是confirm类型的对话框。</p><p>我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p>'
		});
	});

	$('#btn-08').click(function(){
		$.dialog({
			type : 'confirm',
	        onBeforeShow : function(){
	        	alert('显示前执行~~')
	        },
	        onShow : function(){
	        	alert('显示完成后执行~~')
	        },
	        onBeforeClosed : function(){
	        	alert('关闭前执行~~')
	        },
	        onClosed : function(){
	        	alert('关闭后执行，可以看页面title是否改变~~');
	        	document.title = '我已经关闭拉！';
	        },
			contentHtml : '<p>我是confirm类型的对话框。</p><p>我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。我只是用来占位的内容展示，仅仅用来占位撑起提示内容的高度。</p>'
		});
	});


	/*  
	 * info type 
	 */
	$('#btn-09').click(function(){
		$.dialog({
			type : 'info',
	        infoText : '加载中…',
			infoIcon : 'images/icon/loading.gif',
			// autoClose : 2500			
		});
	});

	$('#btn-10').click(function(){
		$.dialog({
			type : 'info',
	        infoText : '操作成功',
			infoIcon : 'images/icon/success.png',
			autoClose : 2500
		});
	});

	$('#btn-11').click(function(){
		$.dialog({
			type : 'info',
	        infoText : '操作失败',
			infoIcon : 'images/icon/fail.png',
			autoClose : 2500			
		});
	});


	$('#btn-12').click(function(){
		$.dialog({
			type : 'info',
	        contentHtml : '<img class="info-icon" src="images/icon/success.png" alt="操作成功" /><p class="info-text">操作成功</p>',
			autoClose : 2500			
		});
	});


	$('#btn-13').click(function(){
		var infoDialog = $.dialog({
			type : 'info',
			infoText : '加载中…',
			infoIcon : 'images/icon/loading.gif'			
		});

		window.setTimeout(function() {
	        infoDialog.dialog.update({
	        	autoClose : 1500,
		    	infoText : '操作成功',
				infoIcon : 'images/icon/success.png'
     		});
     	}, 2500);
	});


	$('#btn-14').click(function(){
	    var infoDialog = $.dialog({
	        type : 'tips',
	        autoClose : 2500,
	        infoText : '正在提交中…',
	        infoIcon : 'images/icon/loading.gif'
	    });    
	});

	$('#btn-15').click(function(){
	    var infoDialog = $.dialog({
	        type : 'tips',
	        autoClose : 2500,
	        infoText : '正在提交中…'
	    });    
	});

	$('#btn-16').click(function(){
	    var infoDialog = $.dialog({
	        type : 'tips',
	        autoClose : 2500,
	        contentHtml : '<img class="info-icon" src="images/icon/loading.gif" alt="正在提交中…"><span class="info-text">正在提交中…</span>'
	    });    
	});


});

js/dialog.js
/*
 * dialog
 * http://sufangyu.github.io
 * 1.0.3(2016-07-15)
 */
;(function(win,$){

    /*
     * Private methods 
     */    
    var wrap, overlay, content, title, close, cancelBtn, okBtn, delBtn, settings, timer, aTimer;

    var _renderDOM = function(){
        if( $('.dialog-wrap').length > 0){
            return;
        }

        clearTimeout(timer);
        clearTimeout(aTimer);
        settings.onBeforeShow();
        
        $('body').append( dialogWrapper = $('<div class="dialog-wrap '+ settings.dialogClass +'"></div>') );
        dialogWrapper.append(
            overlay = $('<div class="dialog-overlay"></div>'),
            content = $('<div class="dialog-content dialog-content-animate"></div>')
        );
        solveTapBug = $('<div class="solve-tap-bug" style="margin:0;padding:0;border:none;background:rgba(255,255,255,0.01); -webkit-tap-highlight-color:rgba(0,0,0,0); width:100%; height:100%; position:fixed; top:0px; left:0px;"></div>').insertBefore(dialogWrapper);
        

        switch (settings.type){
            case 'alert' :
                if(settings.showTitle){
                    content.append(
                        title = $('<div class="dialog-content-hd"><h4 class="dialog-content-title">'+ settings.titleText +'</h4></div>')
                    );
                }
                content.append(
                    contentBd = $('<div class="dialog-content-bd">'+ settings.contentHtml +'</div>')
                );
                content.append(
                    contentFt = $('<div class="dialog-content-ft"></div>')                   
                );
                contentFt.append(
                    okBtn = $('<button class="dialog-btn dialog-btn-ok '+ settings.buttonClass.ok +'" >'+ settings.buttonText.ok +'</button>')
                );
                break;

            case 'confirm' :
                if(settings.showTitle){
                    content.append(
                        title = $('<div class="dialog-content-hd"><h4 class="dialog-content-title">'+ settings.titleText +'</h4></div>')
                    );
                }
                content.append(
                    contentBd = $('<div class="dialog-content-bd">'+ settings.contentHtml +'</div>')
                );
                content.append(
                    contentFt = $('<div class="dialog-content-ft"></div>')
                );
                contentFt.append(
                    cancelBtn = $('<button class="dialog-btn dialog-btn-cancel '+ settings.buttonClass.cancel +'" >'+ settings.buttonText.cancel +'</button>'),
                    okBtn = $('<button class="dialog-btn dialog-btn-ok '+ settings.buttonClass.ok +'" >'+ settings.buttonText.ok +'</button>')
                );
                break;

            case 'info' :
                var infoContent = settings.contentHtml || '<img class="info-icon" src="'+ settings.infoIcon +'" alt="'+ settings.infoText +'" /><p class="info-text">'+ settings.infoText +'</p>';
                content.append(
                    contentBd = $('<div class="dialog-content-bd">'+ infoContent +'</div>')
                );
                dialogWrapper.addClass('dialog-wrap-info');
                content.addClass('dialog-content-info').removeClass('dialog-content-animate');
                break;

            case 'tips' :
                var tipsContent = settings.contentHtml || (settings.infoIcon ? '<img class="info-icon" src="'+ settings.infoIcon +'" alt="'+ settings.infoText +'" />' : '') + '<span class="info-text">'+ settings.infoText +'</span>';
                content.append(
                    contentBd = $('<div class="dialog-content-bd">'+ tipsContent +'</div>')
                );
                dialogWrapper.addClass('dialog-wrap-tips');
                content.addClass('dialog-content-tips').removeClass('dialog-content-animate');
                break;
        }

        setTimeout(function(){            
            dialogWrapper.addClass('dialog-wrap-show');
            settings.onShow();
            _resize();
        }, 20);

        // 解决zepto无法正常获取实际高度造成限制最大高度失效
        setTimeout(function(){
            _setMaxHeight();
        }, 100);

    };

    var _bindEvent = function() {

        touchEvent.tap($(okBtn), function(){
            settings.onClickOk();
            $.dialog.close();
            return false;
        });

        touchEvent.tap($(cancelBtn), function(){
            settings.onClickCancel();
            $.dialog.close();
            return false;
        });

        // overlay clisk hide
        if( settings.overlayClose ){
            touchEvent.tap($(overlay), function(){
                $.dialog.close();
            });
        }

        // auto close, set autoClose and type isn't info
        if( settings.autoClose > 0 ){
            _autoClose();
        }

        // stop body scroll
        $(document).on('touchmove', function(event){
            if( $(dialogWrapper).find($(event.target)).length ){
                return false;
            }else{
                return true;
            }
        });
        _touchScroll();

    };

    var _autoClose = function(){
        clearTimeout(aTimer);
        aTimer = window.setTimeout(function(){
            $.dialog.close();
        }, settings.autoClose);
    };

    var _setMaxHeight = function(){
        var windowHeight = $(window).height();

        $(contentBd).removeAttr('style');
        if( $(content).height() >= windowHeight - 10 ){
            var contentTitleHeight = $(title).height() + parseInt($(title).css('margin-top')) + parseInt($(title).css('margin-bottom')) + parseInt($(title).css('padding-top')) + parseInt($(title).css('padding-bottom'));
            var contentFtHeight = $(contentFt).height() + parseInt($(contentFt).css('margin-top')) + parseInt($(contentFt).css('margin-bottom')) + parseInt($(contentFt).css('padding-top')) + parseInt($(contentFt).css('padding-bottom'));
            var contentBdSpace =  parseInt($(contentBd).css('margin-top')) + parseInt($(contentBd).css('margin-bottom')) + parseInt($(contentBd).css('padding-top')) + parseInt($(contentBd).css('padding-bottom'));
    
            var contentMaxHeight = windowHeight - contentTitleHeight - contentFtHeight - contentBdSpace - 50;
            $(contentBd).css({'max-height': contentMaxHeight, 'overflow-y':'auto'});
        }
    };

    var _resize = function(){

        $(window).on('resize', function(){
            clearTimeout(rTimer);
            var rTimer = window.setTimeout(function(){
                _setMaxHeight();
            },100);
        });
    };

    var _touchScroll = function(){
        var position = {
            x: 0,
            y: 0,
            top: 0,
            left: 0
        };

        // 滚动条top位置 = 原来的滚动条top - 滚动的距离
        $(contentBd).on("touchstart", function(e) {
            position.x = event.changedTouches[0].clientX,
            position.y = event.changedTouches[0].clientY,
            position.top == 0 && (position.top = $(this).scrollTop()),
            position.left == 0 && (position.left = $(this).scrollLeft())
        }).on("touchmove", function(e) {
            $(this).scrollTop(position.top - event.changedTouches[0].clientY + position.y),
            $(this).scrollLeft(position.left - event.changedTouches[0].clientX + position.x)
        }).on("touchend", function(e) {
            if (position.x != 0 || position.y != 0) {
                // return (event.changedTouches[0].clientY - position.y > 20 || event.changedTouches[0].clientX - position.x > 20) && (e = !1),
                position = {
                    x: 0,
                    y: 0,
                    top: 0,
                    left: 0
                }
            }
            e.preventDefault();
        })
    };

    var touchEvent = {
        tap : function(element, fn){           
            if ('ontouchstart' in window || 'ontouchstart' in document) {
                var supportsTouch = true;
            } else if(window.navigator.msPointerEnabled) {
                var supportsTouch = true;
            }

            if(supportsTouch){
                var startTx, startTy;
                element.on('touchstart',function(e){
                    var touches = e.touches ? e.touches[0] : e.originalEvent.touches[0];
                    startTx = touches.clientX;
                    startTy = touches.clientY;
                });
                
                element.on('touchend',function(e){
                    var touches = e.changedTouches ? e.changedTouches[0] : e.originalEvent.changedTouches[0];
                    endTx = touches.clientX,
                    endTy = touches.clientY;
                    // 在部分设备上 touch 事件比较灵敏，导致按下和松开手指时的事件坐标会出现一点点变化
                    if( Math.abs(startTx - endTx) < 6 && Math.abs(startTy - endTy) < 6 ){
                        fn();
                    }
                    e.preventDefault();
                });
            }else{
                element.on('click',function(e){
                    fn();
                });
            }
        }
    };



    /*
     * Public methods 
     */

    $.dialog = function(options) {
        settings = $.extend({}, $.fn.dialog.defaults, options);        
        $.dialog.init();
        return this;
    };   

    $.dialog.init = function(){
        _renderDOM();
        _bindEvent();
    };


    $.dialog.close = function(){
        settings.onBeforeClosed();

        dialogWrapper.removeClass('dialog-wrap-show');
        timer = setTimeout(function(){
            dialogWrapper.remove();
            settings.onClosed();            
        }, 100);

        // cancel stop body scroll
        $(document).on('touchmove', function(event){
            return true;
        });

        // 解决touchend点透，延迟阻止点透层隐藏
        setTimeout(function(){
            solveTapBug.remove();
        }, 350);
    };

    $.dialog.update = function(params) {
        if(params.infoText) {
            content.find('.info-text').html(params.infoText);
        }
        if(params.infoIcon) {            
            content.find('.info-icon').attr('src', params.infoIcon);
        }
        if(params.autoClose>0){
            window.setTimeout(function(){
				if(params.onBeforeClosed) { params.onBeforeClosed(); }
                $.dialog.close();
                if(params.onClosed) { params.onClosed(); }
            }, params.autoClose);
        }
    };


    // 插件
    $.fn.dialog = function(options){
        return this;
    };


    $.fn.dialog.defaults = {
        type : 'alert',     // alert、confirm、info、tips
        titleText : '信息提示',
        showTitle : true,
        contentHtml : '',
        dialogClass : '',
        autoClose : 0,
        overlayClose : false,
        drag : false,

        buttonText : {
            ok : '确定',
            cancel : '取消',
            delete : '删除'
        },
        buttonClass : {
            ok : '',
            cancel : '',
            delete : ''
        },

        infoText : '',      // working in info type
        infoIcon : '',      // working in info type

        onClickOk : function(){},
        onClickCancel : function(){},
        onClickClose : function(){},

        onBeforeShow : function(){},
        onShow : function(){},
        onBeforeClosed : function(){},
        onClosed : function(){}
    }

})(window, window.Zepto || window.jQuery);
